vslib/src/sai_vs_hostintf.cpp

```diff
@@ -29,7 +67,414 @@ sai_status_t vs_send_hostif_packet(
+
+        return SAI_STATUS_FAILURE;
+    }
+
+    // validate SAI_HOSTIF_ATTR_OBJ_ID
+
+    auto attr_obj_id = sai_metadata_get_attr_by_id(SAI_HOSTIF_ATTR_OBJ_ID, attr_count, attr_list);
+
+    if (attr_obj_id == NULL)
+    {
+        SWSS_LOG_ERROR("attr SAI_HOSTIF_ATTR_OBJ_ID was not passed");
+
+        return SAI_STATUS_FAILURE;
+    }
+
+    sai_object_id_t obj_id = attr_obj_id->value.oid;
+
+    sai_object_type_t ot = sai_object_type_query(obj_id);
+
+    if (ot != SAI_OBJECT_TYPE_PORT)
+    {
+        SWSS_LOG_ERROR("SAI_HOSTIF_ATTR_OBJ_ID=%s expected to be PORT but is: %s",
+                sai_serialize_object_id(obj_id).c_str(),
+                sai_serialize_object_type(ot).c_str());
+
+        return SAI_STATUS_FAILURE;
+    }
+
+    // validate SAI_HOSTIF_ATTR_NAME
+
+    auto attr_name = sai_metadata_get_attr_by_id(SAI_HOSTIF_ATTR_NAME, attr_count, attr_list);
+
+    if (attr_name == NULL)
+    {
+        SWSS_LOG_ERROR("attr SAI_HOSTIF_ATTR_NAME was not passed");
+
+        return SAI_STATUS_FAILURE;
+    }
+
+    if (strnlen(attr_name->value.chardata, sizeof(attr_name->value.chardata)) >= MAX_INTERFACE_NAME_LEN)
+    {
+        SWSS_LOG_ERROR("interface name is too long: %.*s", MAX_INTERFACE_NAME_LEN, attr_name->value.chardata);
+
+        return SAI_STATUS_FAILURE;
+    }
+
+    if (strncmp(attr_name->value.chardata, "Ethernet", 8) != 0)
+    {
+        SWSS_LOG_ERROR("interface name should start with EthernetX but is %s", attr_name->value.chardata);
+
+        return SAI_STATUS_FAILURE;
+    }
+
+    std::string name = std::string(attr_name->value.chardata);
+
+    // create TAP device
+
+    SWSS_LOG_INFO("creating hostif %s", name.c_str());
+
+    int tapfd = vs_create_tap_device(name.c_str(), IFF_TAP | IFF_MULTI_QUEUE | IFF_NO_PI);
+
+    if (tapfd < 0)
+    {
+        SWSS_LOG_ERROR("failed to create TAP device for %s", name.c_str());
+
+        return SAI_STATUS_FAILURE;
+    }
+
+    SWSS_LOG_INFO("created TAP device for %s, fd: %d", name.c_str(), tapfd);
+
+    // XXX currently tapfd is ignored, it should be closed on hostif_remove
+    // and we should use it to read/write packets from that interface
+
+    sai_attribute_t attr;
+
+    attr.id = SAI_SWITCH_ATTR_SRC_MAC_ADDRESS;
+
+    sai_status_t status = vs_generic_get(SAI_OBJECT_TYPE_SWITCH, switch_id, 1, &attr);
+
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("failed to get SAI_SWITCH_ATTR_SRC_MAC_ADDRESS on switch %s: %s",
+                sai_serialize_object_id(switch_id).c_str(),
+                sai_serialize_status(status).c_str());
+    }
+
+    int err = vs_set_dev_mac_address(name.c_str(), attr.value.mac);
+
+    if (err < 0)
+    {
+        SWSS_LOG_ERROR("failed to set MAC address %s for %s",
+                sai_serialize_mac(attr.value.mac).c_str(),
+                name.c_str());
+
+        close(tapfd);
+
+        return SAI_STATUS_FAILURE;
+    }
+
+    err = ifup(name.c_str());
```

Comment:
I think orchagent will have logic to bring it up, I am not sure we need to do this in the sai.

Reply:
i brought interface up since forwarding was not possible on interface that was down

Reply:
i removed ifup but i dont see any reason why it can't be brought up at the first place